### PR TITLE
Add label_smoothing to SparseCategoricalCrossentropy

### LINKED ISSUE #23487: Add label_smoothing to SparseCategoricalCrossentropy
## Summary

`label_smoothing` is supported on four of the five crossentropy losses in Keras, but not on `SparseCategoricalCrossentropy`.

| Loss | `label_smoothing` |
| --- | --- |
| `BinaryCrossentropy` | yes |
| `BinaryFocalCrossentropy` | yes |
| `CategoricalCrossentropy` | yes |
| `CategoricalFocalCrossentropy` | yes |
| `SparseCategoricalCrossentropy` | no |

## Motivation

Label smoothing is a standard part of most modern classification and sequence to sequence recipes. The original Transformer paper trains with `0.1`, and the same setting shows up across Inception v3, ViT and the common timm recipes. PyTorch exposes it directly on `nn.CrossEntropyLoss`, which takes integer targets, so anyone porting one of those recipes over expects the sparse loss in Keras to accept it as well.

Today the only way to smooth integer labels is to one-hot encode them first and switch to `CategoricalCrossentropy`. That materializes a target the same shape as `y_pred`. For a segmentation map, or a language model over a large vocabulary, that tensor is already the dominant activation, so doubling it is a real cost for what is conceptually just a different target.

The one-hot target is not actually needed. The smoothed target is

```
(1 - e) * one_hot(y_true) + e / num_classes
```

so the loss splits into the ordinary hard label term plus `e` times the crossentropy against a uniform target. Both pieces can be computed straight from the integer label.

## Proposed API

```python
keras.losses.SparseCategoricalCrossentropy(
    from_logits=False,
    ignore_class=None,
    reduction="sum_over_batch_size",
    axis=-1,
    label_smoothing=0.0,
    name="sparse_categorical_crossentropy",
    dtype=None,
)

keras.losses.sparse_categorical_crossentropy(
    y_true,
    y_pred,
    from_logits=False,
    ignore_class=None,
    axis=-1,
    label_smoothing=0.0,
)
```

The new argument sits after `axis` rather than after `from_logits`, where the other crossentropy losses put it, so that an existing positional call such as `sparse_categorical_crossentropy(y, p, True, 255)` keeps its current meaning. Happy to move it if you would rather the ordering matched the siblings.

The result would match `categorical_crossentropy` on the equivalent one-hot label, and `label_smoothing=0.0` would leave the current output untouched.

I have this working locally and would like to send a PR if the change makes sense.

### PR DESCRIPTION
**Pull request by Neilblaze**
## Description

Adds `label_smoothing` to `keras.losses.SparseCategoricalCrossentropy` and `keras.losses.sparse_categorical_crossentropy`, so integer labels can be smoothed without one-hot encoding them first.

Fixes https://github.com/keras-team/keras/issues/23487 🚀 

### Approach

The smoothed target is `(1 - e) * one_hot(y_true) + e / num_classes`, so the loss is the ordinary hard label term plus `e` times the crossentropy against a uniform target. The second term is evaluated with `ops.categorical_crossentropy` against a uniform tensor, which routes it through the same kernel, clipping and normalization as the categorical path instead of through separate hand written numerics. The one-hot target never enters the graph.

`label_smoothing` is placed after `axis` rather than after `from_logits`. Inserting it earlier would silently reinterpret existing positional calls, so `sparse_categorical_crossentropy(y, p, True, 255)` would read `255` as the smoothing factor with no error. Glad to move it if you would prefer the ordering to match the other crossentropy losses.

Smoothing is applied before the `ignore_class` mask, so ignored entries still contribute exactly zero and the Keras mask is unchanged.

### Testing

New tests in `SparseCategoricalCrossentropyTest` cover

- the config round trip through `get_config` and `from_config`
- equivalence with `categorical_crossentropy` on the one-hot label, for both probabilities and logits, at several smoothing values
- the same equivalence on a segmentation shaped input
- a hand computed value for `label_smoothing=0.3` over three classes
- `label_smoothing=0.0` matching the current output
- the interaction with `ignore_class`, including the mask

Also checked that gradients stay finite for every combination of smoothing value and `ignore_class`, that `float32`, `bfloat16` and `float16` are preserved, and that a compiled fit with `jit_compile=True` followed by a save and load round trip keeps the value.

Ran the `keras/src/losses` suite and the ruff lint and format checks.

<br/>

### AI-assisted contribution disclosure

I used an AI coding assistant during testing and review. I ran the audit across the crossentropy losses that surfaced the gap, wrote the implementation and the tests, and ran the verification described above. I have since read the full diff line by line, & have run the tests locally. I take full responsibility for the correctness of this contribution, and I will be answering review comments myself.

---

## Contributor Agreement

**Please review our [AI-Assisted Contribution Policy](../CONTRIBUTING.md#ai-assisted-contribution-policy) and check all boxes below before submitting your PR for review:**

- [x] I am a human, and not a bot.
- [x] I will be responsible for responding to review comments in a timely manner.
- [x] I will work with the maintainers to push this PR forward until submission.

**Note:** Failing to adhere to this agreement may result in your future PRs no longer being reviewed.

### CODE REVIEW

File: keras/src/losses/losses.py [hunk 1]
```diff
@@ -2330,6 +2341,9 @@ def sparse_categorical_crossentropy(
             considered.
         axis: Defaults to `-1`. The dimension along which the entropy is
             computed.
+        label_smoothing: Float in [0, 1]. If > `0` then smooth the labels. For
+            example, if `0.1`, use `0.1 / num_classes` for non-target labels
+            and `0.9 + 0.1 / num_classes` for target labels.
```
Comment (jeffcarp) [inline keras/src/losses/losses.py]: Can you mention default value here?

File: keras/src/losses/losses.py [hunk 2]
```diff
@@ -2366,6 +2380,29 @@ def sparse_categorical_crossentropy(
         axis=axis,
     )
 
+    if label_smoothing:
```
Comment (jeffcarp) [inline keras/src/losses/losses.py]: I think this should be more explicit `label_smoothing > 0`

File: keras/src/losses/losses.py [hunk 3]
```diff
@@ -2366,6 +2380,20 @@ def sparse_categorical_crossentropy(
...
+        smoothing_axis = canonicalize_axis(axis, len(y_pred.shape))
+        num_classes = ops.cast(ops.shape(y_pred)[smoothing_axis], y_pred.dtype)
+        uniform = ops.divide(ops.ones_like(y_pred), num_classes)
+        uniform_res = ops.categorical_crossentropy(
+            uniform, y_pred, from_logits=from_logits, axis=axis
+        )
+        res = ops.add(
+            ops.multiply(1.0 - label_smoothing, res),
+            ops.multiply(label_smoothing, uniform_res),
+        )
```
[the remark this answers is not in the document]
Comment (jeffcarp) [reply keras/src/losses/losses.py]: Please fix

**Comment by Neilblaze** [comment]
@jeffcarp I've addressed the concerns. PTAL!

**Comment by Neilblaze** [comment]
CC'ing: @jeffcarp , @hertschuh for visibility. Thanks!

**Review by hertschuh — APPROVED** [review]
Thanks!
